core/common/src/main/java/alluxio/util/executor/SerializedTaskRunner.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.util.executor;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class can execute a series of tasks on a separate thread in the order they are submitted.
+ * No guarantees can be made about the timing of how the tasks complete, or what the result of
+ * the tasks are.
+ *
+ * All tasks in the queue are guaranteed to complete before {@link #shutdown(long, TimeUnit)} is
+ * called so long as they complete within the provided timeout. At which time the task runner will
+ * stop accepting new tasks (all calls to {@link #addTask(Runnable)} will return false).
+ *
+ */
+public class SerializedTaskRunner implements Runnable {
```

Comment:
Why do we need this?

Reply:
Each data read handler needs to submit responses using `onNext`, `onError`, and `onCompleted`. In order for the client to understand the protocol properly, we can't have `onNext` messages intermingled with `onError` or `onCompleted`. That's why we used the `SerializingExecutor` before, in order to ensure tasks that were submitted completed one after another.

For one, the previous approach was bad, because it used an internal gRPC class which was not designed to be exposed to users of the library.

Two, because it previously used the same underlying executor as the data reader. In order to send messages back to the client in the proper order, we relied on always being able to submit a task to the serializing executor. If submitting one task failed, the client could time out get a failure because it never actually received the data (failed to submit task which replied with the data), or it could timeout even after the worker attempted to send `onError` or `onCompeted` because it coudl fail to submit the task that calls those methods. Basically when the thread pool filled up before, we would simply stop sending messages back to clients - even ones that had already connected.

This new class is designed so that each new creation of `AbstractReadHandler` corresponds with a lease of a `SerializedTaskRunner`. During the lifetime of the task runner it's guaranteed to be able to run the tasks that are submitted because it uses a dedicated thread. This ensures any client which begins to be handled by the worker can complete or terminate properly because we guarantee that at least one of `StreamObserver#onError` or `StreamObserver#onCompleted` are called.

Reply:
This has been removed